doc/man3/write_log.3
.\"
.\" $Id: write_log.3,v 1.1 2000/07/27 16:59:03 alaffin Exp $
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it would be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" Further, this software is distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
.\" Mountain View, CA  94043, or:
.\"
.\" http://www.sgi.com
.\"
.\" For further information regarding this notice, see:
.\"
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan/
.\"
.TH WRITE_LOG 3 07/25/2000 "Linux Test Project"
.SH NAME
write_log \- a set of routines for logging writes to a history file
.SH SYNOPSIS
.nf
\fB#include "write_log.h"\fP

\fBint wlog_open(struct wlog_file *\fIwfile\fB, int \fItrunc\fB, int \fImode\fB);\fR
\fBint wlog_close(struct wlog_file *\fIwfile\fB);\fR
\fBint wlog_record_write(struct wlog_file *\fIwfile\fB, struct wlog_rec *\fIwrec\fB, long \fIoffset\fB);\fR
\fBint wlog_scan_backward(struct wlog_file *\fIwfile\fB, int \fInrecs\fB, int (*func)(struct wlog_rec *\fIrec\fB), long \fIdata\fB);\fR

\fBchar *Wlog_Error_String;\fR
.fi
.SH DESCRIPTION

The write_log package is a set of routines for creating a history file
of write operations done to a set of files.

It is assumed that the actual pattern written to the file will be
repeated occurrences of a string whose length is no greater than
WLOG_MAX_PATTERN.  See the pattern(3) man
page for routines to conveniently generate buffers of this kind.

wlog_open() initializes the history file contained in wfile->w_file, and
fills in the wfile structure.  If trunc is non-zero, and existing history
file by the same name will be truncated.  If no history file exists, it
will be created with the specified mode.

wlog_close() releases any resources associated with the given wfile.  Use
of wfile after this point is undefined until it is initialized again with
wlog_open().

wlog_record_write() is the main routine for putting a wlog_rec into the
history file.  The caller is responsible for supplying a fully initialized
wlog_rec structure.  If offset is < 0, the record will be appended to the
end of the history file.  If offset is >= 0, the record will be written
at the indicated offset.  This, along with the w_done field in the wlog_rec
structure, provide a mechanism for 'pre-logging' a write, doing the write
operation, and then overlaying the original record with the w_done flag
set to 1.  This is useful for async writes which may not complete in a
timely manner.  It is also useful for seeing which write operations were
pending at the time of a system crash - the ones whose w_done flag is 0 have
not yet been verified as complete.
The return value from wlog_record_write() is the offset
in the history file at which the record was written.

wlog_scan_backward() can be used to conveniently scan a write history file.
The routine scans the file in reverse order (ie. first record written is
scanned last).  For every record found, the user supplied function is called
with 2 parameters:  the read record, and an arbitrary word passed in by the
user.  This word may be interpreted however the user desires.  If nrecs is
greater than 0, up to nrecs will be scanned.  The user supplied function should
return 1 of the following:  WLOG_STOP_SCAN, or WLOG_CONTINUE_SCAN.
WLOG_STOP_SCAN provides a way for the user supplied function to prematurely
abort the scanning process.  WLOG_CONTINUE_SCAN instructs wlog_scan_backward()
to continue scanning the next record.

In order for the history file to be effective, some basic rules must
be followed by the programs using the history mechanism:

.in +.5i
.ll -.5i
The area of the data file being written must be locked from
before the write operation, until after the wlog_record_write()
is complete.  This is necessary to 'synchronize' simultaneous
writes to the same area of a file.  Note that the entire file
does not need to be locked, only the portion being written to.
If the calling program can guarantee that there will never be more
than 1 process writing to the same area of a file at the same time,
locking is not necessary.  (Note:  UNICOS Shared File Systems do not support
record locking.  The whole file is silently locked.)

Pathnames in the history file (w_path field) should be full
pathnames if possible.  This allows validation tools to be
able to find the test files without having to take working
directory considerations into account.
.ll +.5i
.in -.5i

.nf
\fC
.ta .25i +.5i +1i
/*
 * write log file data type.  wlog_open() initializes this structure
 * which is then passed around to the various wlog_xxx routines.
 */

struct wlog_file {
	int	w_afd;	/* append fd */
	int	w_rfd;	/* random-access fd */
	char	w_file[1024];	/* name of the write_log */
};
\fR
.DT
.fi

.nf
\fC
.ta .25i +.5i +2.0i
/*
 * User view of a history file record.  Note that this is not
 * necessarily how the data is formatted on disk (significant
 * compression occurs), so don't expect to od(1) the history file and
 * see things formatted this way.  See the file write_log.h for comments
 * on how the data is actually written to disk.
 */

struct wlog_rec {
	int	w_pid;	/* pid doing the write */
	int	w_offset;	/* file offset */
	int	w_nbytes;	/* # bytes written */
	int	w_oflags;	/* low-order open() flags */
	int	w_done;	/* 1 if io confirmed done */
	int	w_async;	/* 1 if async write (writea) */

	char	w_host[WLOG_MAX_HOST+1];	/* host doing write */
	int	w_hostlen;	/* host name length */
	char	w_path[WLOG_MAX_PATH+1];	/* file written to */
	int	w_pathlen;	/* file name length */
	char	w_pattern[WLOG_MAX_PATTERN+1];	/* pattern written */
	int	w_patternlen;	/* pattern length */
};
\fR
.DT
.fi

Note:  The history files can become very large very quickly if a lot of
processes are logging writes.  This is especially apt to happen if long
pathnames or patterns are used.  This is because the w_host, w_path, and
w_pattern fields are variable length fields when stored on disk.  Thus, use
short pathnames and patterns to minimize the size of the history file.  If
any of the w_path, w_pattern, or w_host fields are not important to you, set
the respective length field to 0 in the wlog_rec structure.

.SH EXAMPLES
This is a simple example of how to initialize a history file, and
record a write to it.

.nf
#include "write_log.h"

main()
{
	struct wlog_rec wrec;
	struct wlog_file wfile;

	...
	strcpy(wfile.w_file, hisfile);
	if (wlog_open(&wfile, 1, 0666) < 0) {
		fprintf("wlog_open failed\n");
		exit(2);
	}

	...

	wrec.w_pid = getpid();
	wrec.w_offset = write_offset;
	wrec.w_nbytes = nbytes;
	wrec.w_oflags = open_flags;
	wrec.w_done = 0;
	wrec.w_async = 0;
	wrec.w_host = 0;             /* don't care about host */

	wrec.w_pathlen = sprintf(wrec.w_path, "%s", path);
	wrec.w_patternlen = sprintf(wrec.w_pattern, "%s", pattern);

	pattern_fill(buf, nbytes, pattern, strlen(pattern), 0);

	... lock fd here ...

	log_offset = wlog_record_write(&wfile, &wrec, -1);
	write(fd, buf, nbytes);
	wrec.w_done = 1;
	wlog_record_write(&wfile, &wrec, log_offset);

	... unlock fd here ...

	...

	/*
	 * Scan the logfile printing records for the file in 'path'.
	 */

	wlog_scan_backward(&wfile, 0, print_log_record, (long)path);
}

int
print_log_record(record, data)
struct wlog_rec	*record;
long		data;
{
	char	*path;

	path = (char *)data;
	if (strcmp(record->w_path, path) == 0) {
		printf("write() of %d bytes to %s at offset %d by pid %d\n",
			record->w_nbytes, record->path, record->w_offset, record->w_pid);
	}

	return WLOG_CONTINUE_SCAN;
}

.fi
.SH "SEE ALSO"
pattern(3).
.SH DIAGNOSTICS
All routines return a value < 0 on failure, and >= 0 on success.  Error
messages can be accessed through Wlog_Error_String.
.SH BUGS
None known.


ChangeLog
LTP-20100831

commit dd59d49d75598d163f1b4599067a9f0dcb70e6c4
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 23:09:49 2010 -0700

    The cacheflush system call does not return EINVAL in the latest Linux
    kernel. see the link:
    A related patch about the cacheflush function:
    http://lkml.org/lkml/2009/4/9/203
    But Ralf had refused this patch:
    http://www.spinics.net/lists/linux-man/msg00906.html
    cacheflush01 testcase checks if the return value is EINVAL when the
    cache argument is not one of ICACHE, BCACHE or DCACHE. So it will fail
    in all boards.
    
    In this modification, checking return value SUCCESS will be added,
    instead of checking EINVAL
    
    Signed-off-by: Feng Gao <feng.gao@windriver.com>

:100644 100644 cce0b70... d0a4b1f... M	testcases/kernel/syscalls/cacheflush/cacheflush01.c

commit ebea45b92a1e30e4480f25291ef4c1ef2eeaa4bf
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 17:14:21 2010 -0700

    Get rid of unnecessary style divergence.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 0ebeca9... b08a2d9... M	testcases/commands/at/at_allow01
:100755 100755 6bca177... 9f9f433... M	testcases/commands/at/at_deny01

commit 08640ca184c1fc1e24e966c3d806a722f3ca2154
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:53:18 2010 -0700

    Add more assertions.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:000000 100644 0000000... 2c590b1... A	testcases/open_posix_testsuite/conformance/interfaces/access/assertions.xml

commit 3cdde2f86af90b3fd4e2b1468a12adfd2a9523f0
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:52:52 2010 -0700

    Note which timer failed in output.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 ac043b0... 8e149ea... M	testcases/open_posix_testsuite/conformance/behavior/timers/1-1.c

commit 207f47c4c9b4783c1a18be32431f5a1b319b9e5a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 19:34:34 2010 -0700

    Update some assertions for raise(2) to SUSv4 reqs.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 38d9542... aa47b7c... M	testcases/open_posix_testsuite/conformance/interfaces/raise/assertions.xml

commit 9c6d2b4cf4f7c166813ddee280ab2bf128eb521c
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:36:51 2010 -0700

    Fix bad syntax that I thought I hadn't committed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 f731d6c... 0ebeca9... M	testcases/commands/at/at_allow01

commit 414747ced26d943a22a3f767deb7f5b4658adb6e
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:32:18 2010 -0700

    Forgot the sleeps.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 9b10030... 6d5eb7f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 a3be617... 60e5241... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c
:100644 100644 34e3afd... abc3279... M	testcases/open_posix_testsuite/conformance/interfaces/pthread_attr_setstack/2-1.c
:100755 100755 01984c5... 536d407... M	testcases/open_posix_testsuite/scripts/generate-makefiles.sh

commit 3222fbecbfb2924a6b901a8e2048c70e1a993018
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:22:07 2010 -0700

    Remove bogus comment.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 c7d81b5... 72a1fcd... M	testcases/open_posix_testsuite/conformance/behavior/timers/2-1.c

commit 7187f0921382bf375cd111b8a480a74ce5786945
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:18:23 2010 -0700

    Repeat robustness work done to 1-1 for 1-2 and 1-3.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 a9d104c... 9b10030... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 af23ebe... a3be617... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c

commit 6c2750a55eaa30fe708c78087688c0d468bfb807
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:12:07 2010 -0700

    Make test more robust.
    
    1. Catch fork() == -1 errors.
    2. Sleep so wait(2) should always succeed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 1efd520... f46ee8f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-1.c

commit 873f6e5b28a6a23367a7117aad436378e56ad09a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:42:53 2010 -0700

    Fix ltp-aiodio.sh script so it execs aio-stress.
    
    Singed-off-by: Jacky Malcles <jacky.malcles@bull.net>

:100755 100755 ed2bce4... ab54c54... M	testscripts/ltp-aiodio.sh

commit 4fd92183833eed128bb6420ece7a6dd45dcd8ec5
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:40:51 2010 -0700

    I've prepared a patch for at_allow01 that essentially duplicates the
    fixes that were checked in for at_deny01. This fixes the script to
    actually run the test when the at utility in stead of returning TCONF,
    correcting test environment, and let the script run normally without
    changing the password. See commit
    a589d12b0b49287d10f43b1a1fb72950819588eb for more details.
    
    Signed-off-by: Henry Yei <hyei@mvista.com>

:100755 100755 cf5b97c... f731d6c... M	testcases/commands/at/at_allow01

commit 217d9b778d275bcbaf5a9fa9e1a5ac2716521cad
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:39:12 2010 -0700

    Chain cmds together to avoid intermediate errors.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 b94bd8f... 6bca177... M	testcases/commands/at/at_deny01

commit 9eeea7939bd73b295ae376dc045e5fbb984d0f38
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Wed Aug 25 20:40:23 2010 -0700

    Fix error message (was hardcoded to SIGALARM).
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>


lib/write_log.c
/*
 * Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * Further, this software is distributed without any warranty that it is
 * free of the rightful claim of any third person regarding infringement
 * or the like.  Any license provided herein, whether implied or
 * otherwise, applies only to this software file.  Patent licenses, if
 * any, provided herein do not apply to combinations of this program with
 * other software, or any other product whatsoever.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
 * Mountain View, CA  94043, or:
 *
 * http://www.sgi.com
 *
 * For further information regarding this notice, see:
 *
 * http://oss.sgi.com/projects/GenInfo/NoticeExplan/
 */
/*
 * This module contains code for logging writes to files, and for
 * perusing the resultant logfile.  The main intent of all this is
 * to provide a 'write history' of a file which can be examined to
 * judge the state of a file (ie. whether it is corrupted or not) based
 * on the write activity.
 *
 * The main abstractions available to the user are the wlog_file, and
 * the wlog_rec.  A wlog_file is a handle encapsulating a write logfile.
 * It is initialized with the wlog_open() function.  This handle is
 * then passed to the various wlog_xxx() functions to provide transparent
 * access to the write logfile.
 *
 * The wlog_rec datatype is a structure which contains all the information
 * about a file write.  Examples include the file name, offset, length,
 * pattern, etc.  In addition there is a bit which is cleared/set based
 * on whether or not the write has been confirmed as complete.  This
 * allows the write logfile to contain information on writes which have
 * been initiated, but not yet completed (as in async io).
 *
 * There is also a function to scan a write logfile in reverse order.
 *
 * NOTE:	For target file analysis based on a write logfile, the
 * 		assumption is made that the file being written to is
 * 		locked from simultaneous access, so that the order of
 * 		write completion is predictable.  This is an issue when
 * 		more than 1 process is trying to write data to the same
 *		target file simultaneously.
 *
 * The history file created is a collection of variable length records
 * described by scruct wlog_rec_disk in write_log.h.  See that module for
 * the layout of the data on disk.
 */

#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/types.h>
#include "write_log.h"

#ifndef BSIZE
#ifdef DEV_BSIZE
#define BSIZE DEV_BSIZE
#else
#define BSIZE BBSIZE
#endif
#endif

#ifndef PATH_MAX
#define PATH_MAX          255
/*#define PATH_MAX pathconf("/", _PC_PATH_MAX)*/
#endif

char	Wlog_Error_String[256];

#if __STDC__
static int	wlog_rec_pack(struct wlog_rec *wrec, char *buf, int flag);
static int	wlog_rec_unpack(struct wlog_rec *wrec, char *buf);
#else
static int	wlog_rec_pack();
static int	wlog_rec_unpack();
#endif

/*
 * Initialize a write logfile.  wfile is a wlog_file structure that has
 * the w_file field filled in.  The rest of the information in the
 * structure is initialized by the routine.
 *
 * The trunc flag is used to indicate whether or not the logfile should
 * be truncated if it currently exists.  If it is non-zero, the file will
 * be truncated, otherwise it will be appended to.
 *
 * The mode argument is the [absolute] mode which the file will be
 * given if it does not exist.  This mode is not affected by your process
 * umask.
 */

int
wlog_open(wfile, trunc, mode)
struct wlog_file	*wfile;
int			trunc;
int			mode;
{
	int	omask, oflags;

	if (trunc)
		trunc = O_TRUNC;

	omask = umask(0);

	/*
	 * Open 1 file descriptor as O_APPEND
	 */

	oflags = O_WRONLY | O_APPEND | O_CREAT | trunc;
	wfile->w_afd =
		open(wfile->w_file, oflags, mode);
	umask(omask);

	if (wfile->w_afd == -1) {
		sprintf(Wlog_Error_String,
			"Could not open write_log - open(%s, %#o, %#o) failed:  %s\n",
			wfile->w_file, oflags, mode, strerror(errno));
		return -1;
	}

	/*
	 * Open the next fd as a random access descriptor
	 */

	oflags = O_RDWR;
	if ((wfile->w_rfd = open(wfile->w_file, oflags)) == -1) {
		sprintf(Wlog_Error_String,
			"Could not open write log - open(%s, %#o) failed:  %s\n",
			wfile->w_file, oflags, strerror(errno));
		close(wfile->w_afd);
		wfile->w_afd = -1;
		return -1;
	}

	return 0;
}

/*
 * Release all resources associated with a wlog_file structure allocated
 * with the wlog_open() call.
 */

int
wlog_close(wfile)
struct wlog_file	*wfile;
{
	close(wfile->w_afd);
	close(wfile->w_rfd);
	return 0;
}

/*
 * Write a wlog_rec structure to a write logfile.  Offset is used to
 * control where the record will be written.  If offset is < 0, the
 * record will be appended to the end of the logfile.  Otherwise, the
 * record which exists at the indicated offset will be overlayed.  This
 * is so that we can record writes which are outstanding (with the w_done
 * bit in wrec cleared), but not completed, and then later update the
 * logfile when the write request completes (as with async io).  When
 * offset is >= 0, only the fixed length portion of the record is
 * rewritten.  See text in write_log.h for details on the format of an
 * on-disk record.
 *
 * The return value of the function is the byte offset in the logfile
 * where the record begins.
 *
 * Note:  It is the callers responsibility to make sure that the offset
 * parameter 'points' to a valid record location when a record is to be
 * overlayed.  This is guarenteed by saving the return value of a previous
 * call to wlog_record_write() which wrote the record to be overlayed.
 *
 * Note2:  The on-disk version of the wlog_rec is MUCH different than
 * the user version.  Don't expect to od the logfile and see data formatted
 * as it is in the wlog_rec structure.  Considerable data packing takes
 * place before the record is written.
 */

int
wlog_record_write(wfile, wrec, offset)
struct wlog_file	*wfile;
struct wlog_rec		*wrec;
long			offset;
{
    int		reclen;
    char	wbuf[WLOG_REC_MAX_SIZE + 2];

    /*
     * If offset is -1, we append the record at the end of file
     *
     * Otherwise, we overlay wrec at the file offset indicated and assume
     * that the caller passed us the correct offset.  We do not record the
     * fname in this case.
     */

    reclen = wlog_rec_pack(wrec, wbuf, (offset < 0));

    if (offset < 0) {
	/*
	 * Since we're writing a complete new record, we must also tack
	 * its length onto the end so that wlog_scan_backward() will work.
	 * Length is asumed to fit into 2 bytes.
	 */

	    wbuf[reclen] = reclen / 256;
	    wbuf[reclen+1] = reclen % 256;
	    reclen += 2;

            if (write(wfile->w_afd, wbuf, reclen) == -1) {
                  sprintf(Wlog_Error_String,
                          "Could not write log - write(%s, %s, %d) failed:  %s\n",
                           wfile->w_file, wbuf, reclen, strerror(errno));
                  return -1;
            } else {
                 offset = lseek(wfile->w_afd, 0, SEEK_CUR) - reclen;
                 if (offset == -1) {
                       sprintf(Wlog_Error_String,
                               "Could not reposition file pointer - lseek(%s, 0, SEEK_CUR) failed:  %s\n",
                                wfile->w_file, strerror(errno));
                       return -1;
                 }
            }
    } else {
            if ((lseek(wfile->w_rfd, offset, SEEK_SET)) == -1) {
                  sprintf(Wlog_Error_String,
                          "Could not reposition file pointer - lseek(%s, %ld, SEEK_SET) failed:  %s\n",
                           wfile->w_file, offset, strerror(errno));
                  return -1;
            } else {
                  if ((write(wfile->w_rfd, wbuf, reclen)) == -1) {
                        sprintf(Wlog_Error_String,
                                "Could not write log - write(%s, %s, %d) failed:  %s\n",
                                 wfile->w_file, wbuf, reclen, strerror(errno));
                        return -1;
                  }
            }
    }

    return offset;
}

/*
 * Function to scan a logfile in reverse order.  Wfile is a valid
 * wlog_file structure initialized by wlog_open().  nrecs is the number
 * of records to scan (all records are scanned if nrecs is 0).  func is
 * a user-supplied function to call for each record found.  The function
 * will be passed a single parameter - a wlog_rec structure .
 */

int
wlog_scan_backward(wfile, nrecs, func, data)
struct wlog_file	*wfile;
int 			nrecs;
int 			(*func)();
long			data;
{
	int		fd, leftover, nbytes, offset, recnum, reclen, rval;
	char   		buf[BSIZE*32], *bufend, *cp, *bufstart;
	char		albuf[WLOG_REC_MAX_SIZE];
	struct wlog_rec	wrec;

	fd = wfile->w_rfd;

	/*
	 * Move to EOF.  offset will always hold the current file offset
	 */

        if ((lseek(fd, 0, SEEK_END)) == -1) {
              sprintf(Wlog_Error_String,
                      "Could not reposition file pointer - lseek(%s, 0, SEEK_END) failed:  %s\n",
                       wfile->w_file, strerror(errno));
              return -1;
        }
	offset = lseek(fd, 0, SEEK_CUR);
        if ((offset == -1)) {
              sprintf(Wlog_Error_String,
                      "Could not reposition file pointer - lseek(%s, 0, SEEK_CUR) failed:  %s\n",
                       wfile->w_file, strerror(errno));
              return -1;
        }

	bufend = buf + sizeof(buf);
	bufstart = buf;

	recnum = 0;
	leftover = 0;
	while ((!nrecs || recnum < nrecs) && offset > 0) {
		/*
		 * Check for beginning of file - if there aren't enough bytes
		 * remaining to fill buf, adjust bufstart.
		 */

		if ((unsigned int)offset + leftover < sizeof(buf)) {
			bufstart = bufend - (offset + leftover);
			offset = 0;
		} else {
			offset -= sizeof(buf) - leftover;
		}

		/*
		 * Move to the proper file offset, and read into buf
		 */
                if ((lseek(fd, offset, SEEK_SET)) ==-1) {
                      sprintf(Wlog_Error_String,
                              "Could not reposition file pointer - lseek(%s, %d, SEEK_SET) failed:  %s\n",
                               wfile->w_file, offset, strerror(errno));
                       return -1;
                }

		nbytes = read(fd, bufstart, bufend - bufstart - leftover);

		if (nbytes == -1) {
			sprintf(Wlog_Error_String,
				"Could not read history file at offset %d - read(%d, %p, %d) failed:  %s\n",
				offset, fd, bufstart,
				(int)(bufend - bufstart - leftover), strerror(errno));
			return -1;
		}

		cp = bufend;
		leftover = 0;

		while (cp >= bufstart) {

			/*
			 * If cp-bufstart is not large enough to hold a piece
			 * of record length information, copy remainder to end
			 * of buf and continue reading the file.
			 */

			if (cp - bufstart < 2) {
				leftover = cp - bufstart;
				memcpy(bufend - leftover, bufstart, leftover);
				break;
			}

			/*
			 * Extract the record length.  We must do it this way
			 * instead of casting cp to an int because cp might
			 * not be word aligned.
			 */

			reclen = (*(cp-2) * 256) + *(cp -1);

			/*
			 * If cp-bufstart isn't large enough to hold a
			 * complete record, plus the length information, copy
			 * the leftover bytes to the end of buf and continue
			 * reading.
			 */

			if (cp - bufstart < reclen + 2) {
				leftover = cp - bufstart;
				memcpy(bufend - leftover, bufstart, leftover);
				break;
			}

			/*
			 * Adjust cp to point at the start of the record.
			 * Copy the record into wbuf so that it is word
			 * aligned and pass the record to the user supplied
			 * function.
			 */

			cp -= reclen + 2;
			memcpy(albuf, cp, reclen);

			wlog_rec_unpack(&wrec, albuf);

			/*
			 * Call the user supplied function -
			 * stop if instructed to.
			 */

			if ((rval = (*func)(&wrec, data)) == WLOG_STOP_SCAN) {
				break;
			}

			recnum++;

			if (nrecs && recnum >= nrecs)
				break;
		}
	}

	return 0;
}

/*
 * The following 2 routines are used to pack and unpack the user
 * visible wlog_rec structure to/from a character buffer which is
 * stored or read from the write logfile.  Any changes to either of
 * these routines must be reflected in the other.
 */

static int
wlog_rec_pack(wrec, buf, flag)
struct wlog_rec	*wrec;
char		*buf;
int             flag;
{
	char			*file, *host, *pattern;
	struct wlog_rec_disk	*wrecd;

	wrecd = (struct wlog_rec_disk *)buf;

	wrecd->w_pid = (uint)wrec->w_pid;
	wrecd->w_offset = (uint)wrec->w_offset;
	wrecd->w_nbytes = (uint)wrec->w_nbytes;
	wrecd->w_oflags = (uint)wrec->w_oflags;
	wrecd->w_done = (uint)wrec->w_done;
	wrecd->w_async = (uint)wrec->w_async;

	wrecd->w_pathlen = (wrec->w_pathlen > 0) ? (uint)wrec->w_pathlen : 0;
	wrecd->w_hostlen = (wrec->w_hostlen > 0) ? (uint)wrec->w_hostlen : 0;
	wrecd->w_patternlen = (wrec->w_patternlen > 0) ? (uint)wrec->w_patternlen : 0;

	/*
	 * If flag is true, we should also pack the variable length parts
	 * of the wlog_rec.  By default, we only pack the fixed length
	 * parts.
	 */

	if (flag) {
		file = buf + sizeof(struct wlog_rec_disk);
		host = file + wrecd->w_pathlen;
		pattern = host + wrecd->w_hostlen;

		if (wrecd->w_pathlen > 0)
			memcpy(file, wrec->w_path, wrecd->w_pathlen);

		if (wrecd->w_hostlen > 0)
			memcpy(host, wrec->w_host, wrecd->w_hostlen);

		if (wrecd->w_patternlen > 0)
			memcpy(pattern, wrec->w_pattern, wrecd->w_patternlen);

		return (sizeof(struct wlog_rec_disk) +
			wrecd->w_pathlen + wrecd->w_hostlen + wrecd->w_patternlen);
	} else {
		return sizeof(struct wlog_rec_disk);
	}
}

static int
wlog_rec_unpack(wrec, buf)
struct wlog_rec	*wrec;
char		*buf;
{
	char			*file, *host, *pattern;
	struct wlog_rec_disk	*wrecd;

	memset((char *)wrec, 0x00, sizeof(struct wlog_rec));
	wrecd = (struct wlog_rec_disk *)buf;

	wrec->w_pid = wrecd->w_pid;
	wrec->w_offset = wrecd->w_offset;
	wrec->w_nbytes = wrecd->w_nbytes;
	wrec->w_oflags = wrecd->w_oflags;
	wrec->w_hostlen = wrecd->w_hostlen;
	wrec->w_pathlen = wrecd->w_pathlen;
	wrec->w_patternlen = wrecd->w_patternlen;
	wrec->w_done = wrecd->w_done;
	wrec->w_async = wrecd->w_async;

	if (wrec->w_pathlen > 0) {
		file = buf + sizeof(struct wlog_rec_disk);
		memcpy(wrec->w_path, file, wrec->w_pathlen);
	}

	if (wrec->w_hostlen > 0) {
		host = buf + sizeof(struct wlog_rec_disk) + wrec->w_pathlen;
		memcpy(wrec->w_host, host, wrec->w_hostlen);
	}

	if (wrec->w_patternlen > 0) {
		pattern = buf + sizeof(struct wlog_rec_disk) +
			wrec->w_pathlen + wrec->w_hostlen;
		memcpy(wrec->w_pattern, pattern, wrec->w_patternlen);
	}

	return 0;
}
